pkg/grpc/server.go

```diff
@@ -83,77 +88,89 @@ func (values watches) cancel() {
 }
 
 func (s *server) process(stream stream, reqCh <-chan *api.DiscoveryRequest, implicitTypeURL string) error {
-	// unique nonce for req-resp pairs; the server ignores stale nonces
-	var nonce int64
+	// unique nonce for req-resp pairs per xDS stream; the server ignores stale nonces.
+	// nonce is only modified within send() function.
+	// nonce 0 is special and always ignored.
+	var streamNonce int64
+	var values watches
+	defer func() {
+		values.cancel()
+	}()
 
 	// sends a response by serializing to protobuf Any
-	send := func(resp cache.Response, typeURL string) error {
+	send := func(resp cache.Response, typeURL string) (int64, error) {
 		resources := make([]*any.Any, len(resp.Resources))
+		streamNonce = streamNonce + 1
```

Comment:
for clarity, I suggest prefixing nonce with the type. Like endpoint-x, cluster-y, etc.

Reply:
bit more work, but easier to debug

Reply:
This is the nonce generator, to make sure nonces are not shared between responses. Are you suggesting having four generators that don't overlap?